coderd/database/pubsub/pubsub.go

```diff
@@ -312,10 +330,20 @@ func (p *pgPubsub) recordReconnect() {
 }
 
 // New creates a new Pubsub implementation using a PostgreSQL connection.
-func New(ctx context.Context, database *sql.DB, connectURL string) (Pubsub, error) {
+func New(ctx context.Context, logger slog.Logger, database *sql.DB, connectURL string) (Pubsub, error) {
 	// Creates a new listener using pq.
 	errCh := make(chan error)
-	listener := pq.NewListener(connectURL, time.Second, time.Minute, func(_ pq.ListenerEventType, err error) {
+	listener := pq.NewListener(connectURL, time.Second, time.Minute, func(t pq.ListenerEventType, err error) {
+		switch t {
+		case pq.ListenerEventConnected:
+			logger.Info(ctx, "pubsub connected to postgres")
+		case pq.ListenerEventDisconnected:
+			logger.Error(ctx, "pubsub disconnected from postgres", slog.Error(err))
```

Comment:
I'd say this is more of an info or warning than an error.

Reply:
That's not a "normal" disconnection when we asked it to disconnect.  It means the connection to postgres dropped due to postgres going down or a network error, and we're going to have to try to reconnect.  In the meantime, we could lose some messages sent during the time we are disconnected.  It's bad news and should be an error.

Reply:
Alright, that makes sense. I didn't think about us potentially losing messages. The disconnect/reconnect I see as expected in any scenario where we're dealing with networking, data loss not so much.

Reply:
This got me thinking, how do we propagate this "potential loss of state"? If someone is viewing the dashboard for instance, and the disconnect happens to coincide with a workspace state update, we need a way to propagate that once connection is re-established. A "simple" re-fetch of the relevant information and send it to the user seems sensible, a diff/change detection if we want to be extreme.

Reply:
I updated the Pubsub interface to include a `SubscribeWithErr` function that informs the subscriber about potential data loss, so that it can resync or punt the error up to a higher layer to retry.  Not everywhere uses it yet.